File: user_app/test/clinical_tool_conformance_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:user_app/features/calculators/data/models/clinical_tool_definition.dart';
import 'package:user_app/features/calculators/domain/clinical_tool_evaluator.dart';

void main() {
  final suite =
      jsonDecode(
            File(
              '../clinical-tools/conformance/v1/runtime-fixtures.json',
            ).readAsStringSync(),
          )
          as Map<String, dynamic>;

  for (final toolValue in suite['tools'] as List<dynamic>) {
    final tool = Map<String, dynamic>.from(toolValue as Map);
    final definition = ClinicalToolDefinition.fromJson(
      Map<String, dynamic>.from(tool['definition'] as Map),
    );
    for (final caseValue in tool['cases'] as List<dynamic>) {
      final fixture = Map<String, dynamic>.from(caseValue as Map);
      test('shared runtime conformance: ${fixture['key']}', () {
        final fixedNow = fixture['fixed_now'] == null
            ? null
            : DateTime.parse(fixture['fixed_now'].toString());
        final evaluator = ClinicalToolEvaluator(fixedNow: fixedNow);
        final input = Map<String, Object?>.from(fixture['input'] as Map);
        if (fixture['expected_error'] != null) {
          expect(
            () => evaluator.evaluate(definition, input),
            throwsA(anything),
          );
          return;
        }
        final result = evaluator.evaluate(definition, input);
        final tolerance =
            (fixture['numeric_tolerance'] as num?)?.toDouble() ?? 0;
        _expectValue(
          result.normalizedInputs,
          Map<String, Object?>.from(
            fixture['expected_normalized_input'] as Map,
          ),
          tolerance,
        );
        _expectValue(
          result.values,
          Map<String, Object?>.from(fixture['expected_outputs'] as Map),
          tolerance,
        );
        expect(
          result.interpretation?.label,
          fixture['expected_interpretation'],
        );
        expect(result.recommendations, fixture['expected_recommendations']);
        expect(
          result.warnings.map((item) => item.text),
          fixture['expected_warnings'],
        );
        if (fixture['expected_checklist'] != null) {
          expect(result.checklist, isNotNull);
          _expectValue(
            result.checklist!.toJson(),
            Map<String, Object?>.from(fixture['expected_checklist'] as Map),
            tolerance,
          );
        }
      });
    }
  }
}

void _expectValue(Object? actual, Object? expected, double tolerance) {
  if (expected is Map) {
    expect(actual, isA<Map>());
    final actualMap = actual! as Map;
    expect(actualMap.keys.toSet(), expected.keys.toSet());
    for (final entry in expected.entries) {
      _expectValue(actualMap[entry.key], entry.value, tolerance);
    }
    return;
  }
  if (expected is num) {
    expect(actual, isA<num>());
    expect(
      (actual! as num).toDouble(),
      closeTo(expected.toDouble(), tolerance),
    );
    return;
  }
  expect(actual, expected);
}

File: dashboard/lib/clinical-tool-conformance.test.ts
import { readFileSync } from "node:fs"
import { join } from "node:path"
import { describe, expect, it } from "vitest"

import type { ClinicalToolDefinition } from "@/services/clinical-tool.service"
import { previewClinicalTool } from "./clinical-tool-evaluator"

type FixtureCase = { key: string; input: Record<string, unknown>; fixed_now?: string; expected_normalized_input?: Record<string, unknown>; expected_outputs?: Record<string, unknown>; expected_interpretation?: string; expected_recommendations?: string[]; expected_warnings?: string[]; expected_checklist?: Record<string, unknown>; expected_error?: string; numeric_tolerance?: number }
type FixtureSuite = { tools: Array<{ definition: ClinicalToolDefinition; cases: FixtureCase[] }> }

const suite = JSON.parse(readFileSync(join(process.cwd(), "..", "clinical-tools", "conformance", "v1", "runtime-fixtures.json"), "utf8")) as FixtureSuite

function expectRecord(actual: Record<string, unknown>, expected: Record<string, unknown>, tolerance = 0) {
  expect(Object.keys(actual).sort()).toEqual(Object.keys(expected).sort())
  for (const [key, value] of Object.entries(expected)) {
    if (typeof value === "number") expect(Number(actual[key])).toBeCloseTo(value, tolerance > 0 ? 6 : 10)
    else expect(actual[key]).toEqual(value)
  }
}

describe("shared clinical-tool runtime conformance", () => {
  for (const tool of suite.tools) for (const fixture of tool.cases) {
    it(fixture.key, () => {
      if (fixture.expected_error) {
        expect(() => previewClinicalTool(tool.definition, fixture.input, { fixedNow: fixture.fixed_now })).toThrow()
        return
      }
      const result = previewClinicalTool(tool.definition, fixture.input, { fixedNow: fixture.fixed_now })
      expectRecord(result.normalizedInputs, fixture.expected_normalized_input!, fixture.numeric_tolerance)
      expectRecord(result.values, fixture.expected_outputs!, fixture.numeric_tolerance)
      expect(result.interpretation).toBe(fixture.expected_interpretation)
      expect(result.recommendations).toEqual(fixture.expected_recommendations!)
      expect(result.warnings).toEqual(fixture.expected_warnings!)
      if (fixture.expected_checklist) expect(result.checklist).toEqual(fixture.expected_checklist)
    })
  }
})
